gunrock/app/bc/bc_enactor.cuh

```diff
@@ -246,124 +232,116 @@ struct BCForwardIterationLoop : public IterationLoopBase
-            // printf("iteration=%d \n", iteration);
-
             VertexT s_label = Load<cub::LOAD_CG>(labels + src);
             VertexT d_label = Load<cub::LOAD_CG>(labels + dest);
 
-            if(iteration == 0) {
+            if (iteration == 0) {
                 return (d_label == s_label + 1);
             } else {
                 if (d_label == s_label + 1) {
-                    if (src == src_node[0]) return true; // !! Is this right?
-                    
+                    if (src == src_node) return true; // !! Is this right? YC: it's right
+
                     ValueT from_sigma = Load<cub::LOAD_CG>(sigmas + src);
                     ValueT to_sigma   = Load<cub::LOAD_CG>(sigmas + dest);
                     ValueT to_delta   = Load<cub::LOAD_CG>(deltas + dest);
                     ValueT result     = from_sigma / to_sigma * (1.0 + to_delta);
-                    
+
                     //Accumulate bc value
-                    {
-                        ValueT old_delta    = atomicAdd(deltas + src, result);
-                        ValueT old_bc_value = atomicAdd(bc_values + src, result);
-                    }
+                    ValueT old_delta    = atomicAdd(deltas + src, result);
+                    ValueT old_bc_value = atomicAdd(bc_values + src, result);
                     return true;
                 } else {
                     return false;
-                }               
+                }
             }
         };
 
         auto filter_op = [
-            labels
+            labels, iteration
         ] __host__ __device__ (
             const VertexT &src, VertexT &dest, const SizeT &edge_id,
             const VertexT &input_item, const SizeT &input_pos,
             SizeT &output_pos) -> bool
         {
-            // printf("backward filter_op");
-            return labels + dest == 0;
+            return labels[dest] == 0;
         };
-        
-        // Call the advance operator, using the advance operation
-        // TODO: modify the operator callers according to algorithmic needs,
-        //       this example only uses an advance + a filter, with
-        //       possible optimization to fuze the two kernels.
-        //       Define more operations (i.e. lambdas) if needed
+
+        frontier.queue_reset = true;
+        auto empty_q = frontier.Next_V_Q();
+        empty_q = NULL;
+        // Call the advance operator, using the advance operation.
+        // BC only uses an advance + a filter, with
+        // possible optimization to fuze the two kernels.
+        GUARD_CU2(cudaStreamSynchronize(oprtr_parameters.stream),
+            "Failed before backward advance");
+
         GUARD_CU(oprtr::Advance<oprtr::OprtrType_V2V>(
-            graph.csr(), frontier.V_Q(), frontier.Next_V_Q(),
+            graph.csr(), frontier.V_Q(), empty_q,
             oprtr_parameters, advance_op, filter_op));
 
+        GUARD_CU2(cudaStreamSynchronize(oprtr_parameters.stream),
```

Comment:
Will we always want this `cudaStreamSynchronize` (and/or the one on line 328)?  Should they be added to the template?

Reply:
The two `cudaStreamSynchronize` should be removed, they are part of the debug code